File: framework/scala/src/main/scala/soda/scala/web/ClassLoaderCache.scala
package soda.scala.web

import java.io.File
import java.net.{URL, URLClassLoader}
import scala.collection.mutable

object ClassLoaderCache {

  private val loaderMap = mutable.Map[String, ClassLoader]()

  def remove(path: String): Unit = this.synchronized {
    loaderMap.remove(path)
  }

  def setupForJar(key: String, jarBytes: Array[Byte]): Unit = this.synchronized {
    val parent = Thread.currentThread().getContextClassLoader
    val loader = new JarFileBytesClassLoader(jarBytes, parent)
    loaderMap(key) = loader
    Logger.info(s"new class loader for $key")
  }

  def getForJar(key: String): ClassLoader = this.synchronized {
    loaderMap(key)
  }

}

File: framework/kotlin/src/main/kotlin/soda/kotlin/web/ClassLoaderCache.kt
package soda.kotlin.web

import java.net.URL
import java.net.URLClassLoader

object ClassLoaderCache {

    private val loaderMap = mutableMapOf<String, ClassLoader>()

//    @Synchronized
//    fun set(path: String) {
//        val parent = Thread.currentThread().contextClassLoader
//        val loader = URLClassLoader(arrayOf(File(path).toURI().toURL()), parent)
//        loaderMap[path] = loader
//        Logger.info("new class loader for $path: $loader")
//    }

    @Synchronized
    fun remove(path: String) {
        loaderMap.remove(path)
    }

    @Synchronized
    fun setupForJar(key: String, jarFile: String) {
        val parent = Thread.currentThread().contextClassLoader
        val loader = URLClassLoader(arrayOf(URL("file:$jarFile")), parent)
        loaderMap[key] = loader
        Logger.info("new class loader for $key: $loader")
    }

    @Synchronized
    fun getForJar(key: String): ClassLoader? {
        return loaderMap[key]
    }

}